### ISSUE TITLE
SyncTestCase.assertNoResult should not eat any eventual result.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6291]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6291|
|Type|enhancement|
|Created|2013-02-01 17:33:14Z|
|Branch|https://github.com/twisted/twisted/tree/assertNoResult-passthru-6291|

Although it may be better to use multiple tests, it is often convenient to assert that a deferred hasn't got a result at one point, and then after some processing does have a specific result:
```python
d = Deferred()
self.assertNoResult(d)
d.callback(True)
self.assertTrue(self.successResultOf(d))
```

Currently, `assertNoResult` turns any success or failure result into a successful `None` result, when the deferred eventually fires. `successResultOf` and `failureResultOf` suffer from the same issue, but the use case there is more tenuous.

<details><summary>Searchable metadata</summary>

```
trac-id__6291 6291
type__enhancement enhancement
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_assertNoResult_passthru_6291 branches/assertNoResult-passthru-6291
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1359739994000000 1359739994000000
changetime__1363636882000000 1363636882000000
version__None None
owner__tom_prince tom.prince
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Here's a case to be careful of:

```python
from twisted.trial.unittest import SynchronousTestCase
from twisted.internet.defer import Deferred
from twisted.python.failure import Failure

class Example(SynchronousTestCase):
    def test_foo(self):
        d = Deferred()
        self.assertNoResult(d)
        d.errback(Failure(Exception()))
```

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

And what should the result of that be? I'd assume the following test case would be successful.

```python
from twisted.trial.unittest import SynchronousTestCase
from twisted.internet.defer import Deferred
from twisted.python.failure import Failure

class Example(SynchronousTestCase):
    def test_foo(self):
        d = Deferred()
        self.assertNoResult(d)
        d.errback(Failure(Exception("HI")))
        f = self.failureResultOf(d)
        self.assertEqual(f.value.message, "HI")
```

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince commented|
|-|-|

Which, on further consideration, suggests that at least `failureResultOf` (and thus `successResultOf`) should consume the result of the deferred.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince commented|
|-|-|

(In [37288]) Branching to assertNoResult-passthru-6291.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince commented|
|-|-|

[build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assertNoResult-passthru-629)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @tomprince|
|-|-|

1. The docstring of test_assertNoResult was changed in a weird way.

2. test_assertNoResultSwallowsImmediateFailure could use asserRaises instead of a try/except.

Looks, please merge once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

There's a typo in the news fragment.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set status to `closed`|
|-|-|

(In [37597]) Merge assertNoResult-passthru-6291: SyncTestCase.assertNoResult doesn't affect the eventual result.

Author: tom.prince
Reviewers: therve
Fixes: [#6291](https://github.com/twisted/twisted/issues/6291)

Although it may be better to use multiple tests, it is often
convenient to assert that a deferred hasn't got a result at
one point, and then after some processing does have a
specific result.
